### PR TITLE
[0.6.x] Backport PR #523 to cast data to iceberg table's pyarrow schema

### PR DESCRIPTION
**Pull request by HonahX**
Backport #523

Depends on #560

### CODE REVIEW

File: pyiceberg/table/__init__.py [hunk 1]
```diff
@@ -1010,6 +1053,13 @@ def append(self, df: pa.Table) -> None:
         if len(self.spec().fields) > 0:
             raise ValueError("Cannot write to partitioned tables")
 
+        from pyiceberg.io.pyarrow import schema_to_pyarrow
+
+        _check_schema_compatible(self.schema(), other_schema=df.schema)
+        # cast if the two schemas are compatible but not equal
+        if schema_to_pyarrow(self.schema()) != df.schema:
+            df = df.cast(schema_to_pyarrow(self.schema()))
```
Comment (HonahX) [inline pyiceberg/table/__init__.py:1061]: Need to use `schema_to_pyarrow` directly here since `schema.as_arrow()` is not available in 0.6.x

**Review by kevinjqliu — COMMENTED** [review]
LGTM! Would it be easier to pull #532 in instead of making a new commit?

**Comment by HonahX** [comment]
@kevinjqliu I did not include that PR since that adds a new API and I think for a patch release we should not add any new functionality. I added that one to 0.7.0 milestone. WDYT?

**Comment by kevinjqliu** [comment]
that makes sense, thanks!

**Pull request merged**